src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractWalletController.cs

```diff
@@ -0,0 +1,337 @@
+
+            if (string.IsNullOrWhiteSpace(address))
+                return ErrorHelpers.BuildErrorResponse(HttpStatusCode.BadRequest, "No address", "No address provided");
+
+            try
+            {
+                var transactionItems = new List<ContractTransactionItem>();
+
+                var account = this.walletManager.GetAccounts(walletName).First();
+
+                // Get a list of all the transactions found in an account (or in a wallet if no account is specified), with the addresses associated with them.
+                IEnumerable<AccountHistory> accountsHistory = this.walletManager.GetHistory(walletName, account.Name);
+
+                // Wallet manager returns only 1 when an account name is specified.
+                AccountHistory accountHistory = accountsHistory.First();
+
+                List<FlatHistory> items = accountHistory.History.OrderByDescending(o => o.Transaction.CreationTime).Where(x=>x.Address.Address == address).ToList();
+
+                // Represents a sublist of transactions associated with receive addresses + a sublist of already spent transactions associated with change addresses.
+                // In effect, we filter out 'change' transactions that are not spent, as we don't want to show these in the history.
+                List<FlatHistory> history = items.Where(t => !t.Address.IsChangeAddress() || (t.Address.IsChangeAddress() && !t.Transaction.IsSpendable())).ToList();
+
+                foreach (FlatHistory item in history)
+                {
+                    TransactionData transaction = item.Transaction;
+
+                    // Record a receive transaction
+                    transactionItems.Add(new ContractTransactionItem
+                    {
+                        Amount = transaction.Amount.ToUnit(MoneyUnit.BTC),
+                        BlockHeight = (uint) transaction.BlockHeight,
+                        Hash = transaction.Id,
+                        Type = ContractTransactionItemType.Received,
+                        To = address
+                    });
+
+                    // Add outgoing transaction details
+                    if (transaction.SpendingDetails != null)
+                    {
+                        // Get if it's an SC transaction
+                        PaymentDetails scPayment = transaction.SpendingDetails.Payments?.FirstOrDefault(x => x.DestinationScriptPubKey.IsSmartContractExec());
+
+                        if (scPayment != null)
+                        {
+                            if (scPayment.DestinationScriptPubKey.IsSmartContractCreate())
+                            {
+                                // Create a record for a Create transaction
+                                var contractAddress = this.addressGenerator.GenerateAddress(transaction.SpendingDetails.TransactionId, 0).ToBase58Address(this.network);
```

Comment:
This will work for now, but feels like we should be getting the address from the state directly rather than deriving it deterministically.